src/Microsoft.DotNet.Build.Tasks/PackageFiles/CodeCoverage.targets

```diff
@@ -50,6 +52,47 @@
     </PropertyGroup>
   </Target>
 
+  <!-- Crossgen assemblies need to be replaced by their respective ILs in order for proper coverage reporting -->
+  <Target Name="CopyIlAssembliesToTestHost"  BeforeTargets="GenerateTestExecutionScripts" Condition="'$(CodeCoverageEnabled)'=='true'">
+    <Message Importance="high" Text="Replacing cross gen assemblies targeted for code coverage with respective IL assemblies." />
+    <ItemGroup>
+      <SelectedIlAssemblies Include="@(_CodeCoverageAssemblies -> '$(NETCoreAppTestSharedFrameworkPath)il/%(Identity).dll')"
+        Condition="Exists('$(NETCoreAppTestSharedFrameworkPath)/il/%(Identity).dll') AND Exists('$(NETCoreAppTestSharedFrameworkPath)/ni/%(Identity).dll')">
+          <NativeImagePath>$(NETCoreAppTestSharedFrameworkPath)%(Identity).dll</NativeImagePath>
+      </SelectedIlAssemblies>
+    </ItemGroup>
+
+    <!-- Delete the files under the test host to break any existing hardlinks -->
+    <Delete Files="@(SelectedIlAssemblies->'%(NativeImagePath)')" />
+
+    <!-- Put IL versions in place -->
+    <Copy SourceFiles="@(SelectedIlAssemblies)" 
```

Comment:
I worry about all this copying back and forth as it is likely to cause us to get into weird states. It only takes one locked dll because of a test not shutting down correctly to corrupt this output. I wonder if it might be better to create a parallel output folder to do the the code coverage run on.

Reply:
I agree with the sentiment, I will do some tries copying everything to a separate folder to see how it goes.

Reply:
https://github.com/dotnet/coreclr/pull/11040#issuecomment-297902475 ... I wish I had found this earlier, but, anyway I opened an issue with CoreClr to see if this is something that can be fixed on the profiler side. I will put this on the back burner until I hear back in https://github.com/dotnet/coreclr/issues/18488

Reply:
(See https://github.com/dotnet/corefx/issues/23588#issuecomment-397832843 for some background info)

For tools doing IL rewrite it is basically required to use a separate folder, but for tools using the Profiler API that shouldn't be necessary. That said for the version of OpenCover (4.6.519) being used in CoreFx we need some workaround since it can't match the proper PDB for native images that follow the convention of shipping .ni.pdb (matching native image) and .pdb (matching IL image) - even if PDBs are moved to a different folder.

In the short run I want to reserve the separate folder for tools doing IL rewrite and workaround OpenCover issues by moving PDB files around: there still be some risky for tools depending on symbols but much smaller than using the assemblies directly. After that I will try to get a change to OpenCover in which it can properly find PDBs for native images, so we can stop moving PDBs around.

Notice that, as before, there is an attempt to ensure expected state for non-coverage runs.